File: transaction-api/src/main/scala/fima/transaction/TransactionController.scala
package fima.transaction

import com.twitter.finatra.http.Controller
import com.twitter.finatra.request.QueryParam
import fima.services.transaction
import fima.services.transaction.TransactionServiceGrpc.TransactionServiceBlockingStub
import fima.services.transactionstatistics.TransactionStatisticsServiceGrpc.TransactionStatisticsServiceBlockingStub
import fima.services.transactionstatistics.{TransactionStatisticsResponse, TransactionsStatisticsRequest}

class TransactionController(transactionService: TransactionServiceBlockingStub,
                            transactionStatisticsService: TransactionStatisticsServiceBlockingStub) extends Controller {

  get("/api/transaction/:id") { request: GetTransactionRequest =>
    val request = transaction.GetTransactionRequest.newBuilder().setId(1).build()
    val r = transactionService.getTransaction(request)

    println(s"DEBUG: ${r.getTransaction.getId}")

    r
  }

  get("/api/transaction/recent") { request: RecentTransactionRequest =>
    val request = transaction.GetRecentTransactionsRequest.newBuilder().setLimit(5).build()
    transactionService.getRecentTransactions(request)
  }

  get("/api/transaction/statistics") { request: TransactionStatisticsRequest =>
    val a: TransactionStatisticsResponse = transactionStatisticsService.getStatistics(TransactionsStatisticsRequest.newBuilder().setName("abc").build())
    StatisticsResponse(a.getTransactions)
  }

  put("/api/transaction") { request: PutTransactionRequest =>
    val request = transaction.InsertTransactionRequest.newBuilder().build()
    transactionService.insertTransaction(request)
  }

  delete("/api/transaction/:id") { request: DeleteTransactionRequest =>
    val request = transaction.DeleteTransactionRequest.newBuilder().build()
    transactionService.deleteTransaction(request)
  }

}

case class GetTransactionRequest(@QueryParam id: Int)

case class TransactionStatisticsRequest()

case class RecentTransactionRequest(@QueryParam limit: Int)

case class PutTransactionRequest(@QueryParam description: String)

case class DeleteTransactionRequest(@QueryParam id: Int)

case class StatisticsResponse(transactions: Int)

File: transaction-api/src/main/kotlin/fima.transaction/transaction/TransactionController.kt
package fima.transaction.transaction

import fima.services.transaction.GetRecentTransactionsRequest
import fima.services.transaction.GetTransactionRequest
import fima.services.transaction.MonthInYear
import fima.services.transaction.TransactionServiceGrpcKt
import fima.services.transaction.TransactionsStatisticsRequest
import fima.services.transactionimport.ImportTransactionsRequest
import fima.services.transactionimport.TransactionImportServiceGrpcKt
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.multipart.MultipartFile
import java.nio.charset.Charset
import java.util.UUID

@RestController
class TransactionController @Autowired constructor(
  private val transactionService: TransactionServiceGrpcKt.TransactionServiceCoroutineStub,
  private val transactionImportService: TransactionImportServiceGrpcKt.TransactionImportServiceCoroutineStub
) {

  @GetMapping("/{id}")
  suspend fun getTransaction(@PathVariable("id") transactionId: UUID): Transaction {
    val request = GetTransactionRequest.newBuilder().setId(transactionId.toString()).build()

    return transactionService.getTransaction(request).transaction.simple()
  }

  @GetMapping("/recent")
  suspend fun getRecentTransactions(@RequestParam("offset") offset: Int, @RequestParam("limit") limit: Int): List<Transaction> {
    val request = GetRecentTransactionsRequest
      .newBuilder()
      .setOffset(offset)
      .setLimit(limit)
      .build()

    return transactionService.getRecentTransactions(request).transactionsList.map { it.simple() }
  }

  @GetMapping("/statistics")
  suspend fun getStatistics(): List<TransactionStatistics> {
    val startOfYear = MonthInYear.newBuilder().setMonth(1).setYear(2018)
    val endOfYear = MonthInYear.newBuilder().setMonth(12).setYear(2018)
    return transactionService
      .getMonthlyStatistics(TransactionsStatisticsRequest.newBuilder().setStartDate(startOfYear).setEndDate(endOfYear).build())
      .monthlyStatisticsList
      .map { it.simple() }
  }

  @PutMapping("/import")
  suspend fun importTransactions(@RequestParam("transactions") transactions: MultipartFile) {
    val request = ImportTransactionsRequest.newBuilder().setTransactions(String(transactions.bytes, Charset.forName("UTF-8"))).build()
    transactionImportService.importTransactions(request)
  }

}